text/0000-explicit-tail-calls.md

```diff
@@ -0,0 +1,574 @@
+In the example, the function would need to be evaluated and **then** the addition would need to take place.
+
+A further restriction is on the function signature of the caller and callee.
+The stack frame layout is based on the calling convention, arguments, as well as return types (the function signature in
+short).
+As the stack frame is to be reused it needs to be similar enough for both functions.
+This requires that the function signature and calling convention of the calling and called function need to match exactly.
+
+Additionally, there is a further restriction on the arguments.
+The stack frame of the calling function is reused, it is essentially cleaned up and the called function takes the space.
+As a result it is not possible to pass references to local variables, neither will the called function "return" to the calling function. So all variables not used as an argument are dropped before the call and no cleanup will be done after the call.
+
+If any of these restrictions are not met when using `become` a compilation error is thrown.
+
+Note that using this feature can make debugging more difficult.
+As `become` causes the stack frame to be reused, debugging context is lost.
+Expect to no longer see any parent functions that used `become` in the stack trace,
+or have access to their variable values while debugging.
+
+<!-- Discuss how this impacts the ability to read, understand, and maintain Rust code. Code is read and modified far more often than written; will the proposed feature make code easier to maintain? -->
+As this feature is strictly opt-in and the `become` keyword is already reserved, this has no impact on existing code.
+
+<!-- If applicable, provide sample error messages, deprecation warnings, or migration guidance. -->
+<!-- (TODO Error messages once an initial implementation exists) -->
+
+<!-- (TODO migration guidance) -->
+
+
+## Teaching
+For new Rust programmers this feature should probably be introduced late into the learning process, it requires
+understanding some advanced concepts and the current use cases are likely to be niche. So it should be taught similarly
+as to programmers that already know Rust.
+
+## Examples
+On to some examples. Starting with how `return` and `become` differ, two example use cases, and some potential
+pitfalls. 
+
+### The difference between `return` and `become`
+[difference]: #difference
+The difference to `return` is that `become` drops function local variables **before** the function call
+instead of after. So the following function ([original example](https://github.com/rust-lang/rfcs/issues/2691#issuecomment-1136728427)):
+```rust
+fn x() {
+    let a = Box::new(());
+    let b = Box::new(());
+    become y(a);
+}
+```
+
+The drops will be elaborated by the compiler like this:
+```rust
+fn x() {
+    let a = Box::new(());
+    let b = Box::new(());
+    drop(b); // `a` is not dropped because it is moved to the callee
```

Comment:
It's unclear to me from the text whether the drop will happen before evaluation of function arguments or not.

Say we have this:
```rust
fn x() {
    let a = Box::new(());
    let b = Box::new(());
    become y(foo(a));
}
```
whether it will be elaborated to
```rust
fn x() {
    let a = Box::new(());
    let b = Box::new(());
    drop(b);
    become y(foo(a));
}
```
or
```rust
fn x() {
    let a = Box::new(());
    let b = Box::new(());
    let arg0 = foo(a);
    drop(b);
    become y(arg0);
}
```
?

From the literal reading of this text it sounds to me that the implication is that drop happens before evaluation of function arguments. And I think this can be very problematic for drop elaboration, as we currently don't have any case where whether a variable will be dropped depends on a later use.

I think we should instead clearly specify the order of evaluation -- the function arguments of the tail call happens first, then it's deemed as end-of-scope of the function, then the tail function call happens.

So essentially we can treat it similar to this alternative function:
```rust
fn x() {
    let (arg0,)= {
        let a = Box::new(());
        let b = Box::new(());
        (foo(a), );
    };
    become y(arg0)
}
```

Reply:
Good point! I'll update the text as you suggest. Though, I'm not sure showing this with a scope is the right way, as it can't really be translated if the `become` is used inside an existing scope. Would the following example be clear enough?

```rust
fn x(_arg: ()) {
    let a = Box::new(());
    let b = Box::new(());

    // Move to temporary variables.
    let function_ptr = bar; // The function pointer could be the result of an expression like: fn_list[fn_idx];
    let arg0 = calc(a);

    // End of the function, all variables not used in the `become` call are dropped, as would be done after a `return`.
    drop(b);
    // `a` has been moved into `calc` and does not need to be dropped.
    drop(_arg);

    // Finally, `become` the called function.
    become function_ptr(arg0);
}
```

Reply:
Just for reference, there is also a more detailed section on borrowchecking: https://github.com/phi-go/rfcs/blob/guaranteed-tco/text/0000-explicit-tail-calls.md#borrowchecking

Reply:
For arbitrarily complex functions, this desugaring would work:
```rust
fn x() {
    let a = Box::new(());
    let b = Box::new(1);
    if C {
        let c = Box::new(());
        become y(foo(a));
    }
    if D {
        become z(bar(b));
    }
    // return ();
}
```
to
```rust
enum XReturn {
    Become1(Box<()>),
    Become2(Box<i32>),
    Return(()),
}
fn x_inner() {
    let a = Box::new(());
    let b = Box::new(1);
    if C {
        let c = Box::new(());
        return XReturn::Become1(foo(a));
    }
    if D {
        return XReturn::Become2(bar(b));
    }
    return XReturn::Return(());
}
fn x() {
    match x_inner() {
        XReturn::Become1(arg0) => become y(arg0),
        XReturn::Become2(arg0) => become z(arg0),
        XReturn::Return(ret) => return ret,
    }
}
```
Essentially [defunctionalization](https://en.wikipedia.org/wiki/Defunctionalization).

Reply:
Thank you for the example, this is an approach I did not think of. Though, as the original example is part of the guide-level explanation, I'm not quite convinced that using the scope example is the best way to teach about the performed drop elaboration. For the implementation there is also the section on borrowchecking which is hopefully already precise enough.

Reply:
I have updated the section by merging it with the borrow checking section, see here: https://github.com/rust-lang/rfcs/pull/3407/commits/0b47449cb4b5ae6b4ecbb7ef19c4e6da650e8b05.